google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlCallable.java

```diff
@@ -0,0 +1,146 @@
+            adjustingIntervalMs);
+    ApiFuture future = innerCallable.futureCall(request, context);
+    future.addListener(flowControllerRunnable, MoreExecutors.directExecutor());
+    return future;
+  }
+
+  class DynamicFlowControlRunnable implements Runnable {
+
+    private final FlowController flowController;
+    private final FlowControlEventStats flowControlEvents;
+    private final DynamicFlowControlStats dynamicFlowControlStats;
+    private final long targetLatency;
+    private final long adjustingIntervalMs;
+    private Stopwatch timer;
+
+    DynamicFlowControlRunnable(
+        @Nonnull FlowController flowController,
+        @Nonnull FlowControlEventStats flowControlStats,
+        @Nonnull DynamicFlowControlStats stats,
+        long targetLatency,
+        long adjustingIntervalMs) {
+      this.flowController = flowController;
+      this.flowControlEvents = flowControlStats;
+      this.dynamicFlowControlStats = stats;
+      this.targetLatency = targetLatency;
+      this.adjustingIntervalMs = adjustingIntervalMs;
+      timer = Stopwatch.createStarted();
+    }
+
+    @Override
+    public void run() {
+      dynamicFlowControlStats.updateLatency(timer.elapsed(TimeUnit.MILLISECONDS));
+      long lastAdjustedTimestamp = dynamicFlowControlStats.getLastAdjustedTimestampMs();
+      long now = System.currentTimeMillis();
+      // Avoid adjusting the thresholds too frequently
+      if (now - lastAdjustedTimestamp < adjustingIntervalMs) {
+        return;
+      }
+      double meanLatency = dynamicFlowControlStats.getMeanLatency();
+      boolean throttled =
+          flowControlEvents.getLastFlowControlEvent() == null
+              ? false
+              : (now - flowControlEvents.getLastFlowControlEvent().getTimestampMs()
+                  <= TimeUnit.MINUTES.toMillis(5));
+      if (meanLatency > targetLatency * 3) {
+        // Decrease at 30% of the maximum
+        decrease(
+            lastAdjustedTimestamp, now, flowController.getMaxOutstandingElementCount() * 3 / 10);
+      } else if (meanLatency > targetLatency * 1.2) {
+        // Decrease at 10% of the maximum
+        decrease(lastAdjustedTimestamp, now, flowController.getMaxOutstandingElementCount() / 10);
+      } else if (throttled && meanLatency < targetLatency * 0.8) {
+        // If latency is low, and there was throttling, then increase the parallelism so that new
+        // calls will not be throttled.
+
+        // Increase parallelism at a slower than we decrease. The lower rate should help the
+        // system maintain stability.
+        increase(
+            lastAdjustedTimestamp, now, flowController.getMaxOutstandingElementCount() * 5 / 100);
+      } else if (throttled
+          && flowController.getCurrentOutstandingElementCount()
+              < flowController.getMaxOutstandingElementCount() * 0.05
+          && meanLatency < 2 * targetLatency) {
+        // When parallelism is reduced latency tends to be artificially higher.
```

Comment:
I haven't tested this yet, but this is similar to what we had in ResourceLimiter